src/blockdata/locktime.rs

```diff
@@ -0,0 +1,595 @@
+///
+/// Bitcoin is able to safely use this value because a block height greater than 500,000,000 would
+/// never occur because it would represent a height in approximately 9500 years. Conversely, block
+/// times under 500,000,000 will never happen because they would represent times before 1986 which
+/// are, for obvious reasons, not useful within the Bitcoin network.
+///
+/// [Bitcoin Core]: https://github.com/bitcoin/bitcoin/blob/9ccaee1d5e2e4b79b0a7c29aadb41b97e4741332/src/script/script.h#L39
+const LOCK_TIME_THRESHOLD: u32 = 500_000_000;
+
+/// A transaction with nLocktime==0 is able to be included immediately in any block.
+pub const ZERO: LockTime = LockTime::Blocks(Height(0));
+
+/// A lock time value, representing either a block height or a UNIX timestamp (seconds since epoch).
+///
+/// Used for transaction lock time (`nLockTime` in Bitcoin Core and [`crate::Transaction::lock_time`]
+/// in this library) and also for the argument to opcode 'OP_CHECKLOCKTIMEVERIFY`.
+///
+/// ### Relevant BIPs
+///
+/// * [BIP-65 OP_CHECKLOCKTIMEVERIFY](https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki)
+/// * [BIP-113 Median time-past as endpoint for lock-time calculations](https://github.com/bitcoin/bips/blob/master/bip-0113.mediawiki)
+///
+/// # Examples
+/// ```no_run
+/// # use bitcoin::blockdata::locktime::LockTime;
+/// # let n = LockTime::from_consensus(100);          // n OP_CHECKLOCKTIMEVERIFY
+/// # let lock_time = LockTime::from_consensus(100);  // nLockTime
+/// # use LockTime::*;
+/// // To compare lock times there are various `is_satisfied_*` methods, you may also use:
+/// let is_satisfied = match (n, lock_time) {
+///     (Blocks(n), Blocks(lock_time)) => n <= lock_time,
+///     (Seconds(n), Seconds(lock_time)) => n <= lock_time,
+///     _ => panic!("handle invalid comparison error"),
+/// };
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub enum LockTime {
+    /// A block height lock time value.
+    ///
+    /// # Examples
+    /// ```rust
+    /// use bitcoin::blockdata::locktime::{LockTime, Height};
+    ///
+    /// let block: u32 = 741521;
+    /// let n = LockTime::from_consensus(block);
+    /// assert!(n.is_block_height());
+    /// assert_eq!(n.to_consensus_u32(), block);
+    /// ```
+    Blocks(Height),
+    /// A UNIX timestamp lock time value.
+    ///
+    /// # Examples
+    /// ```rust
+    /// use bitcoin::blockdata::locktime::{LockTime, Time};
+    ///
+    /// let seconds: u32 = 1653195600; // May 22nd, 5am UTC.
+    /// let n = LockTime::from_consensus(seconds);
+    /// assert!(n.is_block_time());
+    /// assert_eq!(n.to_consensus_u32(), seconds);
+    /// ```
+    Seconds(Time),
+}
+
+impl LockTime {
+    /// Constructs a `LockTime` from an nLockTime value or the argument to OP_CHEKCLOCKTIMEVERIFY.
```

Comment:
In be119a025a8930ad60985a2c2da0975329b945ba:

We should also have `from_height` and `from_time` constructors, and should use thesein all the examples/constants ... `from_consensus` really only ought to be used when starting from a `Transaction::lock_time` field.

Reply:
Yeah, or just `LockTime::from(Height::try_from(741521))`

Reply:
Added `from_height` and `from_time`, including rustdoc examples on usage.